Fix project-local flags being ignored

bairyn opened the pull request:
This was originally the first of two parts to the ‘Track static vs. dynamic dependencies.’ pull request ([#8461](https://github.com/haskell/cabal/pull/8461)), which I've cleaned up and split into two separate PRs.

This PR fixes an issue in which configuration flags such as ‘--disable-library-vanilla’ could be silently dropped when a package was built and installed, and it also adds a test that sets up a scenario to demonstrate the issue.

(I ran the cabal-testsuite/PackageTests test suite and found ‘432 tests, 6 skipped, 0 unexpected passes, 0 unexpected fails’.)

bairyn commented:
I just pushed amendments to this MR that updates the remaining tests (along with a rebase), outside PackageTests (isolated changes: https://github.com/bairyn/cabal/commit/644612052874ab76375ed5321b984d8e57182ed8).

gbaz commented:
Thanks for the thoroughness of this PR and tests, for being responsive to the ask to refactor the big PR, and especially for your patience in explaining it in all the details! All very much appreciated.

gbaz commented:
I'm approving as is, but it would be great to add a changelog entry to the changelog dir as well.

bairyn commented:
Thanks gbaz.

I just updated this MR with the following updates:
- Rebased onto upstream, so it's more up-to-date.
- Added changelog entry.
- Skip the new test on Windows (I noticed the tests I added fail on Windows for some reason; I'll just skip them on Windows for now).

bairyn commented:
No, they're different.  I think what happened in the OS X case is that [the Grinch stole the vowels from the package names](https://github.com/haskell/cabal/issues/7209), so when the test looks for ‘basic-0.1-’ to extract the hash (as it's done in UniqueIPID), it can't find it because it isn't looking for ‘bsc-0.1-’.  I just pushed an update to the hash extraction utility to also check for ‘bsc-0.1-’, so that the utility in the test works also on OS X.

I think what happened in the Windows case is that the installation I set up (with Chocolatey) and the one that the CI server uses don't support dynamic builds.  I set up a VM to reproduce the scenario, and I found that just creating a simple ‘cabal init’ application with ‘--enable-shared --enable-executable-dynamic --disable-library-vanilla’ fails with an error since the dynamic files are missing for the ‘base’ package.  I updated the note around ‘skipIfWindows’.

MR Updates:
- Search for ‘bsc-0.1-’, so that the new test supports OS X.
- Add note about skipping on Windows (needs support for dynamic builds).

Mikolaj reviewed (approved):
Thanks a lot and especially for the test.

Please set one of the merge_me labels and in 2 days it should get auto-merged.

Mikolaj commented:
Hi there! Heads up that we most probably need to improve this PR before we can release cabal 3.10. The main issue is raised in #8721, some more information is in #8723 and both of these link to some other related tickets. Comments welcome. Ideas welcome. Experiments welcome.

phadej commented:
@Mikolaj the main issue is #8719

Mikolaj commented:
https://github.com/haskell/cabal/issues/8719 is the release blocker and we hope it will fall out of https://github.com/haskell/cabal/issues/8721, which is the main task at hand, if the current investigation is accurate. If it's not solved by https://github.com/haskell/cabal/issues/8721, we have to re-asses. Anyway, everything is linked from the two tickets are mentioned.

phadej commented:
#8721 won't solve #8719. The `plan.json` change is only masking the breaking changes done in. The proper fix is #8723, but it essentially reverts what this patch did.

phadej commented:
To be clear: I don't think you can have *this* and fix #8719 at the same time. local tarballs should not be affected by flags meant for local packages.

The fact `v2-install --lib` is done how it is is unfortunate.

There should be a way to explicitly tell that "when you install/build/.. package `foo` apply this flags to package `foo`", that would be generally useful too. (And we can do that in `cabal.project`, I guess - and maybe the that can be done for fake project the `v2-install --lib` builds)

Mikolaj commented:
OK, I stand corrected.

gbaz commented:
After examining this for a while I'm coming around to the point of view of @phadej 

A way to see the issue is that I _think_ that even with this PR, `cabal install foo-pkg --disable-library-vanilla` will ignore that flag if run on a package foo-pkg in a remote repo. Before that pr, this would fail also if run inside a cabal project. After this pr, if its run inside a project directory, it will apply the flag, but if its run on a remote package, it won't. That seems inconsistent. The design, for better or worse, is that cabal _never_ lets you apply flags when running a `v2-install`. I don't think this fixes that underlying issue, it just confuses things further. 

I think the change proposed in https://github.com/haskell/cabal/pull/8723 is the most straightforward (but I don't understand the comment about it breaking `cabal install --lib` ??). We would couple that with flipping the test.

I believe this PR actually packed in another change, which is applying the static and dynamic flags to the setup configs as well, and I think that should stick around. (That was actually the basis on which I had supported this PR to begin with, not realizing it packed in this other thing too!). So that's why I'd propose selective flipping that one line while keeping other refactors (which also seem generally useful), as opposed to a wholesale revert.

Mikolaj commented:
@bairyn: I haven't heard from you and we have to wrap up for the 3.10 (pre-)release. Are you OK with me reverting this PR wholesale on 3.10 and master branches and we can then slowly work our way towards salvaging the clarification and simplification ideas that are contained in it, using the @phadej input? Would that break any of your existing work, e.g., on the other PRs (I hope not, given they compile and test fine on just the bare master branch)?

Mikolaj commented:
The (half-)reversion I'd propose is in #8744.
